Amazon Card Animation

sungik-choi opened the pull request:
<https://sungikchoi.com/amazon-clone>
목데이터는 PR 후 따로 진행해보려 합니다. 모바일은 시간이 된다면...
버튼을 눌러 아이템의 마지막 -> 첫번째로 갈 때 자연스럽게 가지 않는 부분은 더 생각해보고 제작하려 합니다.

sungik-choi committed:
Docs: 파일 추가

Html, Css, Js 파일 추가

sungik-choi committed:
Docs: 파일 경로 수정

폴더 삭제했다.

sungik-choi committed:
Docs: README.md 파일 수정

sungik-choi committed:
Refactor: JS / ES 클래스 형식으로 변경

기존의 코드를 클래스를 나눠 구현했다. 이미지와 텍스트가 들어있는 Carousel 클래스, 이전버튼, 다음버튼이 들어있는 Button 클래스. (Card는 미구현)
캐러셀에서는 스스로 움직이는 기능만 구현하고, 버튼이 클릭됐다는 카운트를 넘겨주는 방식으로 구현했다. 맞는 방식인지는 모르겠다.
일종의 의인화를 한건데,  너무 클래스가 많아지는 것 같기도 하다.

sungik-choi committed:
Refactor: HTML 목업 이미지 추가 / CSS 변경

이미지는 목업 이미지로 변경하고, 카드 컨트롤을 위해 class를 추가했다. CSS는 h1 스타일만 추가했다.

sungik-choi committed:
Add: 카드 클래스 제작

forEach 문을 돌려서 각자의 인덱스만큼 캐러셀이 움직이게 변경했다. 그런데 뭔가 이상하다. 카드 클래스가 있으나마나한 느낌.
월요일에 수업했던 예제 코드를 보며 Carousel 클래스로 Button, Card 모두 모아서 제작해봐야겠다.
클래스 안에 addEventListener도 넣어야겠다.

sungik-choi committed:
Refactor: Carousel 단일 클래스로 합치기

Button, Card 클래스가 하는 일이 너무 적은 거 같아, 캐러셀 클래스 하나로 통합했다. (clickcard(), clickbutton())
문법적으로 이상한 부분들을 수정하고, 클래스 생성시 생성자의 인자를 객체로 활용하는 법을 이제 이해하고 사용해봤다.
변수명을 이해하기 쉽게 다듬었다.
HTML CSS에서 불필요한 부분을 삭제했다.

sungik-choi committed:
Add: 랜덤 노출 기능 추가

아이템 사이즈를 기반으로 난수를 생성하고, 페이지 생성시 slideitem메서드로 위치이동을 시켜주는 기능 추가.

sungik-choi committed:
Docs: JS 파일명 변경 및 Class 콜백함수 변경

버튼 클릭시 카드 활성화 되는 부분 추가하고, 이벤트리스너의 콜백함수들을 클래스 하위 메서드로 나눠서 입력.

sungik-choi committed:
Add: CSS 스타일 추가 및 변경

요즘 유행한다는 뉴모피즘 스타일로 UI CSS 소스 따와서 변경해봤다.
구글 웹폰트 추가.

<https://neumorphism.io/#007bb6>

sungik-choi committed:
Add: 카드 클릭 시 카드 active 되는 기능 추가

문제는 activeCard() 메서드에서, 이미 this.card.forEach 반복문을 돌고 있다는 점인데
반복문 안에 동일한 반복문을 돌고 싶지는 않아서, 37번행 자리에 activeCard() 메서드에서 반복문 블럭안의 내용을 넣어봤지만
그렇게 하면 제대로 작동하지 않고, 반복문을 한번 더 돌아야만 제대로 작동한다.
아직 이 부분을 제대로 이해하지 못하겠다.

sungik-choi committed:
Refactor: 클래스명 일관성있게 변경

__ -> - 으로 모두 변경

sungik-choi committed:
Refactor: 코드리뷰 후 Carousel 인자 수정

기존에 모두 option으로 받았던 부분을 필수 인자로 변경했습니다. index 값이 size를 초과하거나, 0보다 작지 않으면 index 값을 따로 줄 수 있도록 코드를 작성했습니다.
너무 간단하다고 생각했던 index값 변경 로직은 메서드로 정의하지 않고 그대로 콜백함수에 넣었습니다.

sungik-choi committed:
Add / Refactor: 애니메이션 기능 추가 및 리팩토링

옵션 매개변수로 트랜지션 옵션들을 받아 setAnimation() 메서드로 애니메이션을 구현했다. 그런데 새로고침 시에 애니메이션이 일어나는 부분이 마음에 안들고,
사용자 경험에 해를 끼칠 수 있다고 생각했다. slideItem 에서 offset 이라는 키와 밸류값을 리턴하게 해서, 그 값과 아이템의 transform이 일치하면
그 때 transition 스타일을 추가해서 초기 새로고침시에는 트랜지션이 일어나지 않게 했다. 맞는 방식인지는 모르겠다. offset을 생성자에서 선언하기는 싫다보니
slideItem에서 리턴을 하는 방식을 택하게 됐다.

코드 리뷰를 받으며 Index 자체를 옵션값으로 받게 돼서, setIndex()라는 메서드로 인덱스 값을 조작하도록 했다.
인덱스 옵션 값, 랜덤 인덱스 옵션 값을 매개변수로 받는데, 옵션이 없을 경우를 기본 매개변수를 지정하는 방식으로 처리했다. setIndex가 리턴한 Index값이 index가 된다.

최초 실행시에 실행하는 메서드들을 setup() 메서드에 몰아놓았다. 가독성이 향상되었는지? 는 잘 모르겠다. 생성자가 너무 길다고 느꼈기 때문에...
clickButton, clickCard 의 메서드명이 조금 모호하다고 생각해서, 이벤트를 더한다는 의미를 가진 이름으로 메서드명을 변경했다. activeCard도 마찬가지

sungik-choi committed:
Add: 나만의 selector 모듈 파일 작성

쿼리 셀렉터는 $, 셀렉터 all은 $$로 사용할 수 있게 따로 자바스크립트 파일을 작성해서 import 시켰다.
eslint의 규칙이 브라우저 자바스크립트와는 조금 다른듯 하다. 제대로 작동하는 코드에 빨간줄이 쳐있다.

sungik-choi committed:
Merge branch 'sungik-choi' into animation

crongro commented on carousel.js:
```diff
@@ -1,59 +1,81 @@
+import { $, $$ } from "./selector.js";
```
es modules  방식 사용해보는 거 좋아요.
심심하면(?) 자바스크립트  모듈의 역사를 찾아보세요.
흥미진진한 다양한 변화를 볼 수 있답니다.

crongro commented on carousel.js:
```diff
@@ -1,59 +1,81 @@
-    this.button = document.querySelector(option.button);
-    this.card = document.querySelectorAll(option.card);
+    this.index = this.setIndex(option.index, option.useRandomIndex);
+    this.transition = {
```
속성이름이  transition  라고 하니까 css 속성명과 겹쳐서  전 좀  별로인거 같아요.

crongro commented on carousel.js:
```diff
@@ -1,59 +1,81 @@
-    this.clickCard();
+    this.setAnimation(this.slideItem().offset, this.transition.duration, this.transition.timingFunction);
+    this.addEventClickBtn();
+    if (this.card) {
```
this.card 라고만 하니깐  약간 어색하네요. 
자바스크립트에 타입이 없어서 이런게 어렵죠. ㅎㅎ 
bCard라고라도 ?

crongro commented on carousel.js:
```diff
@@ -1,59 +1,81 @@
+    }
+  }
+
+  setIndex(index = 0, randomIndexOption = false) {
```
default parameter 좋죠! 잘했음.

crongro commented on carousel.js:
```diff
@@ -1,59 +1,81 @@
   }
 
-  clickButton() {
+  setAnimation(currentTransformOffset, duration = 0.5, timingFunction = `ease`) {
```
0.5 라는 기본값이 특정메서드에 있는건 좀 숨겨져있는거 같아요.
변경가능성이 높은 값이라 생성자에 위치해도될 거 같고, 
다른 기본값과 함께 묶여서 초기값으로 지정되면  더 좋을 듯 합니다.

crongro commented on selector.js:
```diff
@@ -0,0 +1,9 @@
+function $(element) {
```
selector 와 같은 본인만의 유틸리티를 계속 늘려보시죠.

crongro reviewed (changes requested):
코드 수정이 잘 되고 있는거 같아요.

그런데, 기능은 수정이 필요합니다.
슬라이드가 끝까지 갔을때, 다시 처음으로 되돌아오고 있는데 한방향으로 계속  무한슬라이딩이  되면  좋겠어요.
이거 수정하고 다시  PR보내주세요.

sungik-choi committed:
Add/Refactor: 순환 기능 추가 및 코드 리팩토링

캐러셀 아이템이 순환되게 만들기 위해 item li의 맨 앞과 뒤에 클론된 요소들을 추가했다.
cycleCarousel() 이라는 새로운 메서드를 만들었다. 트랜지션이 끝났을 때, 현재 인덱스가 클론된 요소라면 트랜지션을 삭제하고, 인덱스를 실제(?) 인덱스로 변경한다.
html파일에 요소들이 추가되면서, 기존에 index를 계산하던 방식이 다 어긋나서 이를 리팩토링했다. 불필요해진 메서드는 삭제하거나 병합했다.
클론된 요소가 2개 추가됐으므로, 원래 4개의 아이템이 6개로 늘어나고 -> 그 중 내가 원하는 아이템은 1~4번째의 아이템인걸 고려해서 인덱스 값을 변경했다.
한 가지 걱정되는 부분은 +1 / -1 / 1 숫자가 그대로 코드안에 있는 게... 적절하지 못한 것 같다.
이외에 코드리뷰 받으며 개선할 사항들을 수정했다. 변수명을 변경하고, this.slide 에 기본값을 메서드가 아닌 생성자에 넣었다.

현재 상단 카드의 하이라이트(css) 변경이  트랜지션앤드 이후에 일어나는데, (69행) 불만족스럽다. 기존 클릭 이벤트에 할당됐을때는 자연스러운 느낌이었는데,
트랜지션앤드에 묶이니 자연스럽지 않은 느낌이다. 그렇다고 클릭 이벤트에 묶자니, 맨처음과 맨마지막 클론 요소들이 있는 부분에선 부자연스럽게 동작한다.
또, 버튼을 연속적으로 클릭했을때 부자연스러움이 있다. 45번, 49번행으로 인덱스를 벗어나는 건 막아뒀지만,
아예 트랜지션앤드 전에는 클릭 이벤트가 동작하지 않도록 변경하고 싶다. 그렇게 된다면 굳이 45번, 49번행 코드가 필요하지 않을 것 같다.

crongro commented:
네 무산롤링은 됩니다. 
그런데, 빠르게 누르면 위에 메뉴가 건너띄고 보이지 않는 문제가 있어요.
어색해보이는데 나중에 시간되면 마저 수정해보세요.
